source/bc/mus_bc_passiveScalar_module.fpp

```diff
@@ -204,6 +206,110 @@ contains
+    integer, intent(in) :: neigh(:)
+    !> global time information
+    type(tem_time_type), intent(in)  :: sim_time
+    !> pointer to field variable in the state vector
+    integer, intent(in) :: varPos(:)
+    !> number of Scalars in the scheme var system
+    integer, intent(in) :: nScalars
+    !> scheme variable system
+    type(tem_varSys_type), intent(in) :: varSys
+    !> position of derived quantities in varsys
+    type(mus_derVarPos_type), intent(in) :: derVarPos
+    !> scheme global boundary type
+    type(glob_boundary_type), intent(in) :: globBC
+    !> scheme global boundary type
+    type(mus_physics_type), intent(in) :: physics
+    !> current field
+    integer, intent(in) :: iField
+    !> mixture info
+    type(mus_mixture_type), intent(in) :: mixture
+    ! -------------------------------------------------------------------- !
+    integer :: iElem, iDir, QQ, invDir, elemPos
+    real(kind=rk) :: rhoDef(globBC%nElems(iLevel)) ! Density on boundary element
+    real(kind=rk) :: inv_rho_phy
+    real(kind=rk) :: fTmp( nScalars * globBC%nElems(iLevel) )
+    integer :: posInBuffer, bcPress_pos
+    ! ---------------------------------------------------------------------------
+    
+    QQ = layout%fStencil%QQ
+    inv_rho_phy = 1.0_rk / physics%fac(iLevel)%press * cs2inv
+
+    ! position of boundary pressure in varSys
+    bcPress_pos = me%bc_states%pressure%varPos
+    ! get pressure variable from spacetime function
+    call varSys%method%val(bcPress_pos)%get_valOfIndex( &
+      & varSys  = varSys,                               &
+      & time    = sim_time,                             &
+      & iLevel  = iLevel,                               &
+      & idx     = me%bc_states%pressure                 &
+      &           %pntIndex%indexLvl(iLevel)            &
+      &           %val(1:globBC%nElems(iLevel)),        &
+      & nVals   = globBC%nElems(iLevel),                &
+      & res     = rhoDef                                )
+
+    ! convert physical pressure into LB density
+    rhoDef = rhoDef * inv_rho_phy
+
+    do iElem = 1, globBC%nElems( iLevel )
+      posInBuffer = globBC%elemLvl( iLevel )%posInBcElemBuf%val( iElem )
+      fTmp( (iElem-1)*nScalars+1: (iElem-1)*nScalars+QQ ) &
+        &       = bcBuffer( (posInBuffer-1)*nScalars+varPos(1) : &
+        &                   (posInBuffer-1)*nScalars+varPos(QQ)  )
+    end do
+
+    do iElem = 1, globBC%nElems( iLevel )
+      elemPos = globBC%elemLvl(iLevel)%elem%val( iElem )
+      do iDir = 1, layout%fStencil%QQN
+        if( globBC%elemLvl(iLevel)%bitmask%val(iDir, iElem )) then
+          invDir = layout%fStencil%cxDirInv(iDir)
+
+          state(?FETCH?( iDir, iField, elemPos, QQ, nScalars,nSize, neigh )) = &
+            &   - fTmp( (iElem-1)*nScalars + invDir ) &
+            &   + 2._rk*layout%weight( invDir ) * rhoDef(iElem) 
+
+        end if ! bitMask
+      end do ! iDir
+    end do ! iElem
```

Comment:
Just a suggestion. This element loop could be replaced with a link loop see `velocity_bounceback` in `mus_bc_fluid_module.fpp`

Reply:
Thanks for the suggestion. But in this case, the "out" population is only computed by its inverse direction of population. So, it's different from "velocity_bounceback" and needs to read from the boundary element in any way. I still keep the original code.